tests/functionaltests/functionaltest-api.h

```diff
@@ -385,4 +402,88 @@
 #endif
 #define EXPECT_OBJCNE_MSG(val1, val2, format, ...)                                   \
     FunctionalTestLog::SetUserComment(GTestLogPrivate::Format(format, __VA_ARGS__)); \
-    EXPECT_OBJCNE(val1, val2);
\ No newline at end of file
+    EXPECT_OBJCNE(val1, val2);
+
+// clang-format off
+#pragma push_macro("TEST")
+#define TEST(className, methodName)  \
+    \
+    class className##_##methodName { \
+        TEST_CLASS(className##_##methodName) \
+    \
+        TEST_METHOD(Test); \
+    };  \
+    \
+    void className##_##methodName::Test()
+
+#pragma push_macro("TEST_F")
+#define TEST_F(className, methodName)                  \
+/* This assumes that there already exists a className type with SetUp and TearDown methods to call */ \
+    \
+    class className##_##methodName : className { \
+        TEST_CLASS(className##_##methodName) \
+    \
+        TEST_METHOD(Test); \
+        TEST_METHOD_SETUP(_Setup); \
+        TEST_METHOD_CLEANUP(_Teardown); \
+        \
+        /* TODO: fix up inheritance to not use GTestClasses if not needed. */ \
+        virtual void TestBody() { \
+            Test(); \
+        } \
+    }; \
+    \
+    bool className##_##methodName::_Setup() { \
+        SetUp(); \
+        return true; \
+    } \
+    bool className##_##methodName::_Teardown() { \
+        TearDown(); \
+        return true; \
+    } \
+    void className##_##methodName::Test()
+
+// Don't override TEST_P. Its fine as is because of the needed INSTANTIATE_TEST_CASE_P.
+// TEST_P just sets up a base class with SetUp, TearDown, and TestBody that will be
+// called with the values.
+#pragma push_macro("INSTANTIATE_TEST_CASE_P")
+#define  INSTANTIATE_TEST_CASE_P(prefix, className, generator, ...)                  \
+/* This assumes that there already exists a className type with SetUp and TearDown methods to call */ \
+    \
+    class className##_##prefix { \
+        TEST_CLASS(className##_##prefix) \
+    \
+        TEST_METHOD(Test) { \
+            WEX::Logging::Log::Comment(L"\nStarting parameterized subtests for " L#className L"_" L#prefix L":"); \
+            ::testing::internal::ParamGenerator<className::ParamType> paramGenerator(generator); \
+            for (const auto& metaFactory : ::testing::UnitTest::GetInstance()->parameterized_test_registry(). \
```

Comment:
>for (const auto& metaFactory : ::testing::UnitTest::GetInstance()->parameterized_test_registry() [](start = 12, length = 96)

This is unfortunately super extra double complicated because TEST_P does (className, testName) and INSTANTIATE_TEST_P does (prefix, className) but expects all values to be run over all TEST_P for a matching className. This means we need to know about all classes that derive from className here to iterate over. Luckily the GTEST framework handles this but it feels super icky to reach in like this.